test/e2e/virtual/workspaces/virtual_workspace_test.go

```diff
@@ -298,14 +320,26 @@ func TestWorkspacesVirtualWorkspaces(t *testing.T) {
 			usersKCPArgs, err := framework.Users(users).ArgsForKCP(t)
 			require.NoError(t, err)
 
+			// create port early. We have to hope it is still free when we are ready to start the virtual workspace apiserver.
+			portStr, err := framework.GetFreePort(t)
```

Comment:
(No action required) The port allocation performed by GetFreePort is problematic to me. There is no guarantee that the port we allocate will remain free - a non-test entity not bound by our file-based locking scheme could easily end up using our chosen port before our process using it starts. Ideally anything relying on this strategy for port allocation would at least retry.

Even better - a process would accept a port of 0 and enable discovery of the port that it gets allocated (e.g. by writing it to a known location on disk, kubeconfig, etc). I always wondered why kube didn't support this directly.

Reply:
the apiserver library allows that in-process. You can instantiate a listener manually with kernel port zero, which means to choose and bind one. Then you can do whatever you want and then start up the server with that listener. This obviously does not work out-of-process.

Reply:
I'm saying it *should* work out-of-process, provided a way to communicate the port e.g. via fs.